enhancements/cluster-logging/forwarding-json-structured-logs.md

```diff
@@ -123,132 +115,128 @@ Two new top-level fields added to the record [data model][data_model]:
 Relationship between `structured` and `message` fields:
 
 * EXACTLY ONE of `message` or `structured` MUST be present.
-* `schema` is OPTIONAL, defined and interpreted by the user.
-* The presence or absence of the `structured` and `message` fields can be used to detect structured/unstructured records.
-
-With an *Elasticsearch* store, `schema` is copied to the existing `viaq_index_name` field to set the index.
-`viaq_index_name` will continue to be used for compatibility and upgrade, but will eventually be removed.
-
-#### ClusterLogForwarder configuration
+* The presence of `structured` or `message` fields identifies structured/unstructured records.
+* The `schema` field is optional.
+  It is provided to represent the schema of a structured record, but its use is up to the consumer.
 
-The elements of the CLF API are:
+For the default *Elasticsearch* store, `schema` is used to generate the index name.
 
-- *Input*: Selects logs, adds meta-data, creates _internal_ records.
-- *Pipeline*: Transforms, filters and routes internal records.
-- *Output*: Connects to consumers, encodes and sends _output_ records.
+For the initial implementation the forwarder will use it to generate the `viaq_index_name` field.
+Future versions of the Elasticsearch store may use `schema` directly.
 
-New fields:
+#### ClusterLogForwarder configuration
 
-`input.stuctured`: (string, default "JSON") Attempt to parse structured entries.
-The only legal value is "JSON", others may be added in future.
+New *pipeline* fields:
 
-An input record contains:
+* `stuctured`: (string, optional) Legal values are "JSON" or "json" (there may be others in future).
+   If set, attempt to parse log entries as JSON objects.
+* `schemaKey`: (string, optional) Use meta-data key as `schema` value if present.
+  - `kubernetes.namespaceName`: Use the namespace name as the schema name.
```

Comment:
@alanconway what are the use cases for introducing the `schemaKey` field? not sure i understand correctly to which of the scenarios this one specifically applies.

Specifically, how does this correlate with the `namespace` key for a pipeline to only consider logs from a particular namespace?

Reply:
The input selectors (log type, namespace, eventually pod labels) are statically configured to apply a user-chosen schema name to an arbitrary set of containers. If you want to add a new schema, you have to change the forwarder config. If you have a lot of labels, your config will be very big.
The schemaKey dynamically applies a metadata value as the schema. You can define a *single* input with schemaKey=namespace and the namespace will be used as the schema. No need to change the CRD every time someone adds a namespace.
IMO the most likely and cleanest use of this whole setup is to create a single CRD with `schemaKey=kubernetes.label.MySchemaLabel`. That means you can just apply MySchemaLabel=whatever to your pods to have whatever labels you want at run-time, no need to mess with the CRD when new schema are defined.
If we don't have schemaKey, then we absolutely need  label selectors, based on the AMDocs/Avaloq discussoins there's really no hope that namespaces will be sufficient. schemaKey is much easier to implement and easier to use.